### ISSUE TITLE
Issue #16781: `NoWhitespaceBefore` blind spot resulting in high failure rate

### ISSUE THREAD

**Issue by Pankraz76** [checkstyle/checkstyle#16781]
### Understanding the `NoWhitespaceBefore` Check

I may be misunderstanding the `NoWhitespaceBefore` check, but my current understanding is that it should detect and flag any unwanted whitespace that appears before specific tokens, including:

- Dots (`.`)
- Semicolons (`;`)
- String literals (`""`)
- Method call parentheses (`()`)
- Other similar syntax elements

The check's purpose is to identify and report any unnecessary whitespace preceding these tokens.

**Testing Observations:**
When testing with 10 straightforward examples, only 1 violation was detected. This low detection rate raises concerns about either:
1. The check's current implementation
2. My interpretation of its intended behavior

### Test Case Demonstration

```java
public void testSpaceViolation2() {
     "".equals(""); // Expected violation
    "" .equals(""); // Expected violation - found by current impl. of NoWhitespaceBeforeCheck
    "". equals(""); // Expected violation
    "".equals (""); // Expected violation
    "".equals( ""); // Expected violation
    "".equals("" ); // Expected violation
    "".equals("") ; // Expected violation
    " ".equals("");
    "".equals(" ");
}
```

<img width="1426" alt="Image" src="https://github.com/user-attachments/assets/19dccb91-67d2-434a-b805-a8a3a688e804" />

**Comment by nrmancuso** [comment]
Please follow the issue template; unit tests can change, and screenshots are not generally helpful when trying to reproduce issues.

**Comment by pankratz76** [comment]
Please see dedicated test PR: #16810

I don´t understand the template stuff yet.

Issue #16781: `NoWhitespaceBefore` blind spot resulting in high failure rate

1 out of 15 is 6.667%, so more then 90% blind-spot is pretty much not to fix this in #16810. Please help me getting this done.

<img width="1286" alt="image" src="https://github.com/user-attachments/assets/19ad8aee-dd4d-4e6b-9cac-13cdf7d15582" />

**Comment by pankratz76** [comment]
#16810 contains all the tests needed to fix. Kindly request your feedback nick on how to proceed here. Maybe #16780 is a good approach if you could help me make it run. Thanks

**Comment by pankratz76** [comment]
is this any good ? #16820

**Comment by nrmancuso** [comment]
> is this any good ? [#16820](https://github.com/checkstyle/checkstyle/issues/16820)

No, this is not. It seems like we are expending a lot of effort here for what should be a simple request to fulfill: update this issue, following the template, and clearly showing the configuration, java file, and a minimal example that highlights the issue, with the latest, released version of checkstyle executed on the code. There should be some comments like you’ve done above where existing and false negatives occur.

This makes it easy for others to reproduce, and everyone to be able to agree on the scope of work to be done and the expectations. The issue description should be the source of truth for the work to be done. No one should have to sift through multiple issues, pull requests, and comments to decipher what the expectations are.

If what I can draw from the existing issue description is accurate, it seems like a widespread issue, and we will likely need to break this up into smaller parts.

**Comment by pankratz76** [comment]
> need to break this up into smaller parts

No, you just need to checkout or copy and recreate my test (in a few seconds - much faster than talking here over and over again without any outcome) and fix it.

Or simply help me getting my solution to work as there is one tiny issue to work on this topic:

https://github.com/checkstyle/checkstyle/pull/16780#issuecomment-2789039341

Please, how to fix? `TreeWalker is not allowed as a parent of`

**Comment by pankratz76** [comment]
I'm really wondering if this is the right way to appreciate and deal with customer feedback provided. At rewrite, we try to make things as easy as possible without enforcing strict overhead template stuff like here. A simple test case like in PR: https://github.com/openrewrite/rewrite-static-analysis/pull/495 is enough for everyone to recreate the issue.

When talking about the tool version, it seems to me that there is a lack of understanding here, as I'm running on the latest dev version as we all do when developing locally. So this looks to me more like an excuse not to deal with critical issues like this.

**Comment by pankratz76** [comment]
As I'm pointing out a 90%+ failure rate, it's pretty clear that the software implementation is flawed and the tests are too, as we surely have 100% test coverage but don't even find 10% of issues in real world - at least not when I'm in charge and control, as demonstrated in numerous PRs and comments that are simple to recreate yourself.

But if there is a lack of courage to acknowledge errors and weaknesses on our own, I'm wondering how we should improve ourselves if not by going weak (accept the issue) and getting stronger (fix the test).

**Comment by pankratz76** [comment]
Here we go again: https://github.com/checkstyle/checkstyle/pull/16823

It's all fitting on one screen. If you still can't reproduce and acknowledge this, then it's clearly an emotional/personal issue that needs someone else to handle it. 

I'm done wasting time explaining how bad this checker is. It's wasting your time too since we're getting nowhere. Let's stop with the excuses and fake "process" - we both know this is about avoiding the real problem.

I'm a deal-breaker - I go straight to the point. Tell me now: are you actually interested in fixing this software? If yes, we'll work on it. If not, someone else will - the information is all there.

The screenshot and CI prove it works. Just recreate it and fix the silly test.

It's not rocket science - just add the few lines and make it work. Stop pretending this can't be reproduced when I keep providing solutions.

If you won't do it, SAY SO. You could've made a template by now - it's obvious. We're just going in circles. I'm here to get shit done. Join me or get out of the way.

<img width="1719" alt="Image" src="https://github.com/user-attachments/assets/71cc1f61-feea-4e0f-86a7-f4c637e2bb09" />

**Comment by pankratz76** [comment]
If you still insist on templates and can't help me just this once by doing it yourself, then SAY SO CLEARLY so we can find someone actually capable of fixing this.

I know we can find someone competent enough to make this work - as we are not ATM.

**Comment by nrmancuso** [comment]
@pankratz76 

**Please keep your tone respectful.** We are all volunteers here in this community, and abrasive or sarcastic comments are generally unproductive; they will not help to move this forward.

The issue template exists to save everyone's time (including yours). Whatever urgency you feel around this doesn’t exempt you from following it. If you're serious about resolving this, I ask that you follow the process.

Once you've done that, I would be happy to take a closer look and help to move things forward.

**Comment by romani** [comment]
I insist on template. I don't understand concern.

>but my current understanding is that it should detect and flag any unwanted whitespace that appears before specific tokens

It is too crazy, to be blind in validation. This Check is very old, and unfortunately very broad in design.

**Comment by pankratz76** [comment]
Please provide the template by checking out my example or show me once so I can learn from it.

I've thought deeply about this and believe we can fix the 95% blind spot in this problematic code, including tests. It's shocking how far it is from its actual purpose, and makes me questioning everything even more, but **we all improve through failure**.

Why is template process better than a working PR? Checking a branch and running tests seems faster than template processing. A failing test is all we need (#16823 proves this).

Sorry for frustration - I felt bullied by process over solutions. The tests (the "judge") prove my case. Fixing these tests will fix implementation.

The 95% failure rate suggests test issues - 5% might be acceptable, but not 95% in a high quality project like this beauty.

We want the same goal, even if we disagree on approach.

Pick yourself just one line from the example and it will fail instantly:

Please don't act like you don't have opportunity to recreate it yourself in seconds. I have tried it like crazy to provide you everything needed.

At rewrite we work exact like this, without any issue:
https://github.com/openrewrite/rewrite-static-analysis/pull/495

`A test is all we need.`

So please let us work this out - it's way too serious to keep posting instead of coding.

I'm not saying you shouldn't use templates, but be a little bit open and flexible in mind to additionally support ready-to-use PRs containing running examples.

``` java
         System.out.println(""); // violation
        System .out.println(""); // violation
        System. out.println(""); // violation
        System.out .println(""); // violation
        System.out. println(""); // violation
        System.out. println(""); // violation
        System.out.println (""); // violation
         "".equals(""); // violation
        "" .equals(""); // violation
        "". equals(""); // violation
        "".equals (""); // violation
        "".equals( ""); // violation
        "".equals("" ); // violation
        "".equals("") ; // violation
```

**Comment by pankratz76** [comment]
or here: second example: #16810

<img width="1733" alt="Image" src="https://github.com/user-attachments/assets/043626da-7459-4eb2-a8aa-5abdd96d1cd9" />


```
missing (41)  : 25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 49, 50, 51, 62, 63, 64, 65, 66, 67, 78, 79, 80, 91, 92, 101, 110, 122, 123, 134, 135, 145, 156, 177, 189, 201, 213, 227
unexpected (1): 161
---
expected      : [25, 26, 27, 28, 29, 30, 31, 32, 33, 34, 35, 36, 37, 38, 39, 49, 50, 51, 62, 63, 64, 65, 66, 67, 78, 79, 80, 91, 92, 101, 110, 122, 123, 134, 135, 145, 156, 177, 189, 201, 213, 227, 238, 249, 260, 269]
but was       : [39, 161, 238, 249, 260, 269]
```

<img width="378" alt="Image" src="https://github.com/user-attachments/assets/568bb8a3-02b0-4303-9a90-6dffe85c2b6b" />

**Comment by pankratz76** [comment]
> This Check is very old, and unfortunately very broad in design.

#16780

We might need a breaking change migrating from `Treewalker` to `FileParser`.  
`AbstractFileSetCheck` "is not allowed below this" - if we proceed this way (which seems doable for this case), it's gonna be a breaking change.

I figured it out myself but didn't make it sufficient yet.  
The implementation works but the setup doesn't, as it's still somehow pretending to be a sub-element even though I've already moved it from `Treewalker`.

Parsing the files might be necessary to detect `whitespace` violations,  
since the modeling layer represents logic that's already left  
the dimension we need to check.  

When adding all tokens, I couldn't find the source code in its raw form  
needed for violation checks.

**Comment by nrmancuso** [comment]
@pankratz76

It’s good that you’re starting to consider user-facing concerns like breaking changes; that kind of discussion is important and is the exact reason that we insist on a clear issue description and approval before work starts.

That said, the tone of your replies comes across as entitled and dismissive. This is a volunteer-driven project, and respectful collaboration is key.

I’m stepping away from this issue now.

**Comment by pankratz76** [comment]
> That said, the tone of your replies comes across as entitled and dismissive. This is a volunteer-driven project, and respectful collaboration is key.

true, even this is not my intend.

Please excuse me. Im easily overpacing.

**Comment by Pankraz76** [comment]
**Subject:** Urgent: Process Over Pragmatism - Let's Align on Priorities

We're facing a 90% blindspot in production, yet we're stuck debating template formalities. This makes no sense. I'm actively trying to fix critical issues, but the process is so slow and rigid that it feels impossible to make progress.

The reality:
- The template requirements are harder than implementing the actual feature
- I've had to fork tools like Checkstyle just to work around these hurdles
- Meanwhile, production issues keep piling up

I'll admit I'm not perfect at following every rule, but the current process isn't helping either. Let's team up and:
1. **Acknowledge this isn't working** - The overhead is disproportionate
2. **Fast-track small fixes** - Approve with minor tweaks
3. **Focus on what matters** - Production stability over paperwork

This isn't about ignoring standards - it's about not letting perfect become the enemy of good. I'm happy to improve, but we need to be practical about solving real problems.

Let's find a way forward today.

**Comment by Pankraz76** [comment]
still broken yes, but dont seem to care that much.

**Comment by Pankraz76** [comment]
- https://github.com/pmd/pmd/blob/ad71cb5a4a670bdb9d6b53d90476d0510dc14618/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/internal/JavaAstProcessor.java#L154
- https://github.com/pmd/pmd/blob/ad71cb5a4a670bdb9d6b53d90476d0510dc14618/pmd-java/src/main/java/net/sourceforge/pmd/lang/java/internal/JavaAstProcessor.java#L161C43-L161C44

**Comment by Pankraz76** [comment]
no need to waste effort onto checkstyle anymore.

Pretenting to have 100% quality and not passing prone and PMD is a very good and hilarious joke, but not some real considerable software.
